### PR TITLE
Add support for Chroma

### PR DESCRIPTION
**Pull request by Trojaner**
**What currently works:**
* Loading Chroma in diffusers format
  * Example: https://huggingface.co/imnotednamode/Chroma-v36-dc-diffusers
  * The repo above also contains a conversion script in case someone wants to use newer versions.
* Generating images 
* TAESD
* prob some other generic stuff

**What is added but not testet yet:**
* Teacache (I have added teacache support by patching Chroma's forward)
* Support for standard SD.Next quantization
  * Includes support for NF4
  * It is really important to skip the distilled_guidance_layer when quantitizing Chroma.
    * https://github.com/huggingface/diffusers/pull/11698#issuecomment-2969717180
    * Quote: "I'm guessing the distilled layer has that range which NF4 doesn't play well with that you are talking about"
    * The linked comment also contains a comparison. The difference in quality is huge!
    * In fact even more layers are recommended to skip:
      * https://huggingface.co/Clybius/Chroma-fp8-scaled/discussions/8 
* Offloading
* Interposing
* Remote VAE
* FasterCache (it should work better than for Flux because Chroma supports conditional generations)
* Extracting lora from checkpoint
* Lora support
  * Chroma is very compatible with Flux loras out of the box.
    * However anything related to a trained to CLIP will become useless as Chroma does not use CLIP.
    * Otherwise Flux loras still work great despite the huge amount of training Chroma already has gone through.
    * Since Chroma has some architectural changes like removed or added layers, some keys will not map when using Flux loras
      * These missing keys must be ignored in all SD.Next lora loading methods. Despite the missing keys the lora will still work. 
      * This can be confirmed with Comfy, it just prints a lot of warnings about missing layers but the lora still work.
    * Loras that target the T5 text encoder of FLUX (text_encoder_2) need to be patched and redirected to text_encoder instead as this is where Chroma's T5 text encoder is located instead.

**What probably could be added:**
* IP Adapter
  * Diffusers supports it but first results with models trained for Flux were not that great apparently.
  * Can be still integrated so people can use it as soon someone trains it for Chroma.
  * See https://github.com/huggingface/diffusers/pull/11725#discussion_r2152593248
* CFGZero (needs it's own pipeline)
* Para attention (needs it's own pipeline)
* DifferentialDiffusion
* PuLID

**Current issues:**
* Support for single file models is not fully implemented yet.
* Cannot generate images in same quality as Comfy and it has become worse in new diffusers PRs:
Prompt: "A cheerful collage showcases 4 individual leopard geckos, each in a separate square. The geckos vary in color and pattern, but all are shown close-up and looking directly at the camera. Backgrounds differ, illustrating various terrarium elements and simple surfaces. High-quality shots, bright lighting, amateur photo quality. The overall aesthetic is heartwarming and simple."
Negative: "NSFW"
Resolution: 1024x1024
CFG: 4.7
Seed: 249
Steps: 26
Sampler: Euler FlowMatch
Scheduler: betas

Reference Image from Chroma v10 [see related post](https://civitai.com/images/61592306)
image

Clostest I managed to get before PR https://github.com/huggingface/diffusers/pull/11725 and https://github.com/huggingface/diffusers/pull/11729:
image

After merging these PR's into a diffusers local branch and padding prompt_embeds and negative_prompt_embeds:
image

* Native prompt attention does not follow prompt like it should. Here is an example for the same image:

image

**What is blocked:**
* ChromaImg2ImgPipeline does not exist yet. It be added with https://github.com/huggingface/diffusers/pull/11725
* Currently we need to call `pad_prompt_embeds_to_same_size` in `get_weighted_text_embeddings_chroma` for xhinker. This restriction will be fixed too once https://github.com/huggingface/diffusers/pull/11729 get's updated.
  * https://github.com/huggingface/diffusers/pull/11725#issuecomment-2982180434 
* nunchaku does not support Chroma yet. They are waiting also waiting for the same PRs.
  * https://github.com/mit-han-lab/nunchaku/issues/167
* Chroma does not officially support ControlNet yet.
  * Chroma author recommends trying to merge ControlNet into Chroma. 
  * https://huggingface.co/lodestones/Chroma/discussions/27#6826dd4d0da9f976835d47d6
**Things to consider:**
* The diffusers commit hash is taken from the still WIP PR https://github.com/huggingface/diffusers/pull/11725. It needs to be updated once the PR gets fixed and merged.
  * This PR includes some of the changes from https://github.com/huggingface/diffusers/pull/11729
*  I have added Chroma as a new model type "chroma" instead of defining it like FLUX as "f1".
  * The main reason for this is that there are simply too many differences between them regarding what they support.
  * Some examples:
     * Flux supports InfiniteYou while Chroma does not
     * Flux supports Para-Attention while Chroma does not
     * Flux supports ControlNet but Chroma does not
     * Chroma supports full FasterCache while Flux does not
     * Chroma supports negative prompts while Flux does not
     * etc.
  * It would be very difficult to handle these cases if Chroma was defined as f1 instead of its own model type.
  * The people who integrated Chroma into diffusers also decided to separate Chroma into its own pipeline instead of just patching it into FluxPipeline
    * https://github.com/huggingface/diffusers/pull/11566#issuecomment-2957549110

### CODE REVIEW

**Comment by Ednaordinary** [comment]
> After merging these PR's into a local diffusers branch and having to do padding on prompt_embeds and negative_prompt_embeds, the quality decreased a lot (especially bottom right):

If you can't get batched cond + uncond to work seamlessly then you can use just use the follow up PR for now while testing, as they are equivalent in terms of output quality (your padding may not be though, unsure)

**Comment by hameerabbasi** [comment]
Are we passing in an attention mask explicitly for the prompt? If yes, we should make sure it includes one pad token at the end, but not more than one. That's how default Chroma does its masking.

**Comment by Trojaner** [comment]
> If you can't get batched cond + uncond to work seamlessly then you can use just use the follow up PR for now while testing, as they are equivalent in terms of output quality (your padding may not be though, unsure)

Thanks. I will test it and share the results in case I can confirm a regression. My padding may be wrong as well, not sure I padded with the correct data.

**Comment by Trojaner** [comment]
> Are we passing in an attention mask explicitly for the prompt? If yes, we should make sure it includes one pad token at the end, but not more than one. That's how default Chroma does its masking.

No, only prompt_embeds and negative_prompt_embeds.

**Comment by hameerabbasi** [comment]
> No, only prompt_embeds and negative_prompt_embeds.

Yes, but when calculating those; one pad token needs to be added to the prompt (or equivalently, the unmasked tokens need to include exactly one pad token at the end). They need to be calculated from the prompt somehow. 😉

**Comment by iwr-redmond** [comment]
>  I am a bit confused on this because logs mention TAESD but it doesn't even have FLUX/Chroma compatible models?

There is a Flux TAESD [here](https://huggingface.co/madebyollin/taef1).

**Comment by Trojaner** [comment]
Added prompt_attention_mask for positive and negative, honestly it didn't make the difference I expected.

xhinker prompt attention
image

native / diffusers prompt attention
image

diffusers is now closer to generating 4 images but still not following prompt as it should.

args passed to pipe:
```
'prompt_embeds': tensor([[[-0.0439,  0.1562, -0.0515,  ..., ]]], device='cuda:0', dtype=torch.bfloat16), 
'prompt_attention_mask': tensor([[1., 1., 1., 1., 1., ...]], device='cuda:0', dtype=torch.bfloat16), 
'negative_prompt_embeds': tensor([[[ 4.2480e-02,  3.9291e-04,  3.6377e-02,  ...,  ]]], device='cuda:0', dtype=torch.bfloat16),
'negative_prompt_attention_mask': tensor([[1., 1., 1., 1., ...]], device='cuda:0', dtype=torch.bfloat16),
'guidance_scale': 4.7,
'generator': [<torch._C.Generator object at 0x7d3530a72d90>],
'callback_on_step_end': <function diffusers_callback at 0x7d32bf538860>,
'callback_on_step_end_tensor_inputs': ['latents', 'prompt_embeds', 'noise_pred'],
'num_inference_steps': 26,
'output_type': 'latent',
'width': 1024,
'height': 1024
```

See https://github.com/Trojaner/sdnext/blob/feature/chroma-support/modules/prompt_parser_diffusers.py#L543-#L545 for how I get embeds and attention masks from the pipe:

```py
if "Chroma" in pipe.__class__.__name__: # does not use clip and has no pooled embeds
    prompt_embeds, _, prompt_attention_mask, negative_prompt_embeds, _, negative_prompt_attention_mask = pipe.encode_prompt(prompt=prompt, negative_prompt=neg_prompt, device=device, num_images_per_prompt=1)
    return prompt_embeds, None, prompt_attention_mask, negative_prompt_embeds, None, negative_prompt_attention_mask
```

**Comment by Trojaner** [comment]
@hameerabbasi @Ednaordinary any ideas?

**Comment by Trojaner** [comment]
results with updated xhinker padding
image

**Comment by Trojaner** [comment]
with latest xhinker padding changes:

prompt tokens: 82 unmasked tokens (including 1 pad token + 1 eos token)  + 0 masked tokens = 82

prompt (same with and without mask applied)
> A cheerful collage showcases 4 individual leopard geckos, each in a separate square. The geckos vary in color and pattern, but all are shown close-up and looking directly at the camera. Backgrounds differ, illustrating various terrarium elements and simple surfaces. High-quality shots, bright lighting, amateur photo quality. The overall aesthetic is heartwarming and simple.\<pad>\</s>

negative prompt tokens: 5 unmasked tokens (including 1 pad token + 1 eos token)  + 77 masked padding tokens = 82

negative prompt with mask applied
>NSFW\<pad>\</s>

negative prompt without mask applied
>NSFW\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\<pad>\</s>

image

I think this seems good enough now? To me this looks better than what Comfy generates.

**Comment by vladmandic** [comment]
this is a massive and extremely complete pr

- all the files in `configs/chroma` are only to avoid model config lookup via http  
  typically those are added only for most popular models, not on day-one  
  not sure if we need this for chroma on day-one, but since you already added it, we can keep them  
- `modules/model_chroma_nf4.py`  
  this is based on flux_nf4.py which is really old nf4 loader, not sure if we need it for chroma  
  is there a prequantized nf4 model you really want to use?  
  i'd much rather prefer to use on-the-fly quantization that we have in sdnext  
- in `modules/processing_args.py`, why remove `if prompt_parser_diffusers.embedder is not None`?
- `get_weighted_text_embeddings_chroma` is a massive function  
  for future, it may be easier to add new models without this and have chroma just use build-it encoder  
  and then follow-up later with a pr that adds sdnext long-prompt/attention support  
  but hey, i'm not complaining, this looks really good  
- `modules/teacache/teacache_chroma.py` same as above  
  this is better as a follow-up we don't have a huge pr all-at-once?
- `modules/model_chroma.py` this is core  
  honestly, i'd rather merge and then see what needs to be done  

now, if you want to really be complete about it, how about adding chroma to `html/reference.json` and create 512px thumbnail for it in `models/Reference`?

all-in-all, once merge conflicts are resolved and you say pr is good-to-go, i'm ok with merging and we can continue the work after the merge.

**Comment by Disty0** [comment]
> in modules/processing_args.py, why remove if prompt_parser_diffusers.embedder is not None?

It is redundant. I also removed it when i was adding support for hidream but decided to keep it.
image

**Comment by Trojaner** [comment]
Thanks for the review @vladmandic 

> all the files in configs/chroma are only to avoid model config lookup via http
typically those are added only for most popular models, not on day-one
not sure if we need this for chroma on day-one, but since you already added it, we can keep them

Chroma wasn't in diffusers format yet when I added this (at least not in official repo), but I agree we should still keep them.

> modules/model_chroma_nf4.py
this is based on flux_nf4.py which is really old nf4 loader, not sure if we need it for chroma
is there a prequantized nf4 model you really want to use?
i'd much rather prefer to use on-the-fly quantization that we have in sdnext

Honestly I just took it from FLUX since it didn't look like much effort anyway. Not sure if Chroma needs this, since it's already smaller than FLUX anyway. Hence there might be no need at all for it.

> in modules/processing_args.py, why remove if prompt_parser_diffusers.embedder is not None?

As Disty mentioned, it's already checked in the parent if and hence redunant

> get_weighted_text_embeddings_chroma is a massive function
for future, it may be easier to add new models without this and have chroma just use build-it encoder
and then follow-up later with a pr that adds sdnext long-prompt/attention support
but hey, i'm not complaining, this looks really good

This is actually one of the main features of this PR. The reason is that Chroma uses a unique method with prompt attention masks.  See [here](https://huggingface.co/lodestones/Chroma#tldr-masking-t5-padding-tokens-enhanced-fidelity-and-increased-stability-during-training) for technical details.

In short, based on inpections of the implemenitation code and comparison of same generation results, neither diffusers nor Comfy seem to encode it correctly, which results in a huge quality degradation. This is where SD.Next could really shine, since it might be the only relevant Web UI for now that gets this correctly and that can fully utilize Chroma. Of course this is only until others fix this as well. 

<details>
  <summary>Comparison</summary>
> A cheerful collage showcases 4 individual leopard geckos, each in a separate square. The geckos vary in color and pattern, but all are shown close-up and looking directly at the camera. Backgrounds differ, illustrating various terrarium elements and simple surfaces. High-quality shots, bright lighting, amateur photo quality. The overall aesthetic is heartwarming and simple.


diffusers implementation
image

my implementation
image

Notice how diffusers is not following the prompt correctly, which specifies exactly 4 geckos.
</details>

There is some stuff that might be still missing / wrong. The Chroma documentation mentions a bos_token, but the T5 encoder doesn't have one. It also mentions to keep exactly one pad token unmasked, regardless of if there are none needed. But it  does not mention which one to keep unmasked, e.g. the first one or the last one?

> now, if you want to really be complete about it, how about adding chroma to html/reference.json and create 512px thumbnail for it in models/Reference?

Sure. But keep in mind that Chroma does not update it's diffusers weights while the single file weights are still being updated every 3-4 days. The diffusers weights are 2 versions behind at the moment. See [here](https://huggingface.co/lodestones/Chroma/tree/main).


> modules/teacache/teacache_chroma.py same as above
this is better as a follow-up we don't have a huge pr all-at-once?

Since Teacache is already tested and confirmed to be working, I think it should stay in the PR. 
However I could separate any further work (like support for FLUX loras, single file models, etc.) 

> all-in-all, once merge conflicts are resolved and you say pr is good-to-go, i'm ok with merging and we can continue the work after the merge.

Works for me👍

**Comment by vladmandic** [comment]
> Chroma wasn't in diffusers format yet when I added this (at least not in official repo), but I agree we should still keep them.

👍

> Honestly I just took it from FLUX since it didn't look like much effort anyway. Not sure if Chroma needs this, since it's already smaller than FLUX anyway. Hence there might be no need at all for it.

Lets remove `modules/model_chroma_nf4.py`. That was a separate Flux loader for a very specific NF4 variant.

> As Disty mentioned, it's already checked in the parent if and hence redunant

👍

> This is actually one of the main features of this PR. The reason is that Chroma uses a unique method with prompt attention masks. See [here](https://huggingface.co/lodestones/Chroma#tldr-masking-t5-padding-tokens-enhanced-fidelity-and-increased-stability-during-training) for technical details.This is actually one of the main features of this PR. The reason is that Chroma uses a unique method with prompt attention masks. See [here](https://huggingface.co/lodestones/Chroma#tldr-masking-t5-padding-tokens-enhanced-fidelity-and-increased-stability-during-training) for technical details.

I understand that, I'm just saying that for future PRs we may have 1 that adds base support quicky and then follow-up with such advanced work.

Btw, you may use this section if you want to add tunable params? <https://github.com/vladmandic/sdnext/blob/2714d29993a06c3a542eadbf577a77a6cf5d77dc/modules/shared.py#L417>

> Sure. But keep in mind that Chroma does not update it's diffusers weights while the single file weights are still being updated every 3-4 days. The diffusers weights are 2 versions behind at the moment. See [here](https://huggingface.co/lodestones/Chroma/tree/main).Sure. But keep in mind that Chroma does not update it's diffusers weights while the single file weights are still being updated every 3-4 days. The diffusers weights are 2 versions behind at the moment. See [here](https://huggingface.co/lodestones/Chroma/tree/main).

Unfortunate.

> Since Teacache is already tested and confirmed to be working, I think it should stay in the PR. However I could separate any further work (like support for FLUX loras, single file models, etc.)

That was more of a general comment. Anything that already works there is no point of splitting/removing.
Anything you are ready-to-go with, you can include. I'm just saying that there is no need to delay the PR with advanced work since it can be followed-up after the merge.

**Comment by Trojaner** [comment]
> Btw, you may use this section if you want to add tunable params?

Sure. We can add min padding length and unmask first / unmask last padding token as user configurable params.

> I understand that, I'm just saying that for future PRs we may have 1 that adds base support quicky and then follow-up with such advanced work.

Will do it like this next time 👍

> That was more of a general comment. Anything that already works there is no point of splitting/removing.
Anything you are ready-to-go with, you can include. I'm just saying that there is no need to delay the PR with advanced work since it can be followed-up after the merge.

In that case this PR can be merged to dev at its current state. 
Just need to remove modules/model_chroma_nf4.py or it can be also done after merge instead
